anndata/_config.py

```diff
@@ -0,0 +1,290 @@
+"""
+
+
+class SettingsManager:
+    _registered_options: dict[str, RegisteredOption] = {}
+    _deprecated_options: dict[str, DeprecatedOption] = {}
+    _config: dict[str, object] = {}
+    __doc_tmpl__: str = _docstring
+
+    def describe(
+        self,
+        option: str | Iterable[str] | None = None,
+        *,
+        print_description: bool = True,
+    ) -> str:
+        """Print and/or return a (string) description of the option(s).
+
+        Parameters
+        ----------
+        option
+            Option(s) to be described, by default None (i.e., do all option)
+        print_description
+            Whether or not to print the description in addition to returning it., by default True
+
+        Returns
+        -------
+        The description.
+        """
+        if option is None:
+            return self.describe(
+                self._registered_options.keys(), print_description=print_description
+            )
+        if isinstance(option, Iterable) and not isinstance(option, str):
+            return "\n".join(
+                [self.describe(k, print_description=print_description) for k in option]
+            )
+        registered_option = self._registered_options[option]
+        doc = registered_option.doc.rstrip("\n")
+        if option in self._deprecated_options:
+            opt = self._deprecated_options[option]
+            if opt.message is not None:
+                doc += " *" + opt.message
+            doc += f" {option} will be removed in {opt.removal_version}.*"
+        if print_description:
+            print(doc)
+        return doc
+
+    def deprecate(
+        self, option: str, removal_version: str, message: str | None = None
+    ) -> None:
+        """Deprecate options with a message at a version.
+
+        Parameters
+        ----------
+        option
+            Which option should be deprecated.
+        removal_version
+            The version targeted for removal.
+        message
+            A custom message.
+        """
+        self._deprecated_options[option] = DeprecatedOption(
+            option, message, removal_version
+        )
+
+    def register(
+        self,
+        option: str,
+        default_value: T,
+        description: str,
+        validator: Callable[[T], None],
+        option_type: object | None = None,
+    ) -> None:
+        """Register an option so it can be set/described etc. by end-users
+
+        Parameters
+        ----------
+        option
+            Option to be set.
+        default_value
+            Default value with which to set the option.
+        description
+            Description to be used in the docstring.
+        validator
+            A function which asserts that the option's value is valid.
+        option
+            Optional override for the option type to be displayed.  Otherwise `type(default_value)`.
+        """
+        validator(default_value)
```

Comment:
How about you just do something like this?

```suggestion
        try:
            validator(default_value)
        except (ValueError, TypeError) as e:
            e.add_note(f"in option {option}")
```

Then you can just do `validator=bool` instead of having to craft the message in each validator function

PS: we have some helper to add to `__notes__` in Python 3.9 I think